### ISSUE TITLE
Unity 5 Compability

### ISSUE THREAD

**Issue by CzlowiekDrzewo** [ata4/disunity#77]
Hello!
I'm not sure if this is the right place to talk about it, but do you plan on making DisUnity compatible with Unity 5?

**Comment by Hirotaro** [comment]
It seems that Unity 5 has been released :)

**Comment by ata4** [comment]
Yep, I've noticed and I'll download it. :)

But there are still some issues left to fix before I can support Unity 5. I still have to fix a off-by-one error in the rewritten I/O library.

**Comment by v00d00m4n** [comment]
New game Republique on steam uses Unity 5, assets definetly changed format.
Anyone needs sample files?
Game is very cheap now on steam, btw.

**Comment by ata4** [comment]
I've got an example project for Unity 5 already, which should hopefully enough to figure out the format.

**Comment by zhhispig** [comment]
Thank ata4,I hope i will extract Ori and The Blind Forest Soon,I like the game so much!

**Comment by kR105** [comment]
I'm watching this. Please let me know if you need help reverse engineering the new file format.

**Comment by ata4** [comment]
I guess I could use some help. It looks like the .pdb files don't come with structure information in Unity 5, so it's not that easy this time.

**Comment by grimreefer24601** [comment]
I work on a somewhat competing product, Unity Asset Editor, that I've just got Unity 5 ready with some help from DerPopo over at 7 Days to Die.  Here's what he told me.  Hopefully it helps you out.

These should be all the changes :

    version 13/14 : An additional byte of padding has been added to the TypeTree (after the 4-byte attribute int after the unity version string).
    version 13/14 : TypeField changed completely (except classId). If the classId is smaller than 0 (signed), it consists of 8 DWORDs, otherwise of 4 (the last 4 are the same for both).
    version 13/14 : There is no TypeTree so a TypeField ends after these DWORDS.
    version 14 : There is no more 4 bytes of padding between TypeTree and the assets file table.
    version 14 : There now is a 4-byte boundary alignment after the file count of the assets file table.
    version 14 : The file index for each assets file info now is 8 bytes long.
    version 11-14 : The last 4-byte int (which seemed to be a copy of the file type) in each assets file info is now split into two 2-byte ints (for each assets file info).

**Comment by ata4** [comment]
Thanks, that should help for the start.

**Comment by grimreefer24601** [comment]
No problem! Actually thank you!  I've looked at Disunity for a lot of things.  I'm glad to help you out anyway I can.  I've figured out some more about it, but like you said that is a start, and if you're like me you dig figuring it out..  I'm happy to help in any other way I can.

**Comment by Tristian** [comment]
@grimreefer24601 You wouldn't mind sharing your latest version would you? Best I could find was Beta 3. I'm digging the built in hex viewer etc.

**Comment by DerPopo** [comment]
Unity's current Pro version (5.0.0p4) has a newer file format version than the current Free version (5.0.0f4).
The current file version (0x0F) has only one change in the asset file table.
Each file entry now has an additional byte at the end so it's 25 bytes long. Because the file position is aligned to 4 bytes before each file entry, all entries except the last one take 28 bytes.

If you want to read the dependency list, there has been something new before it (after the asset file table) since version 0x0B.
The first DWORD specifies the amount of items in the list (it looks like some kind of asset reference table).
Each item contains a 4-byte field (which probably is the fileId) and another field (which probably is the pathId) that is 4 bytes long before version 0x0E and 8 bytes long starting with 0x0E.

**Comment by ata4** [comment]
Okay, thanks DerPopo and grimreefer24601, I think I got most parts of the new format now. Just the seemingly random 16/32 bytes in the base class structure are confusing me a bit. Maybe these are UUIDs to identify the class type?

**Comment by SimTheVoid** [comment]
Hello! 
What about version 5.0.1f ?

P.S. Do you accept BTC donations for that great project?

**Comment by ata4** [comment]
I'm still working on the deserializer for Unity 5. No idea yet whether the file format has changed in 5.0.1 or not.

And yeah, I have a Bitcoin address if you like to donate:

1PSbKbJSRMbm5UPuTkWcLLXeUbNXF4vhw6

**Comment by DerPopo** [comment]
The tools I wrote for 5.0.0p3 (file format 0x0F) still work on 5.0.1f1 assets files so I think the format hasn't been changed.

**Comment by ata4** [comment]
Oooh, I found another program database file, webplayer_development_x86.pdb, which actually does contain symbols! That should make things easier. I think I'll update the wiki with the new information first before adding more Unity 5 code.

**Comment by ata4** [comment]
Also, I could use more information about which file version is used in which Unity version. Right now, there's a gap between 9 and 14. I guess these are used in Unity 5 beta versions, but I don't have any files to confirm that.

**Comment by DerPopo** [comment]
The latest Unity 4 version (4.6.4f1) still uses the file format 9 and the first one of Unity 5 I saw (a beta version that had no free version) used 14.
I guess that they created many versions during the internal development and I don't think there are any released games that use these.

**Comment by ata4** [comment]
Yep, I had a look at an older demo project for Unity 5 and found some files that are using version 12 from an internal release 4.9.9a1, so I guess the other versions are for 4.8/4.7 or something. The issue tracker lists 5.0.0a1 as the oldest 5 release, so it should be safe to start with 14.
